fzf-tab.zsh

```diff
@@ -80,19 +97,33 @@ _fzf_tab_compadd() {
     _opts+=("${(@kv)apre}" "${(@kv)hpre}" $isfile)
     __tmp_value+=$'\0args\0'${(pj:\1:)_opts}
 
+
     # dscr - the string to show to users
+    # example dscr:
+    # Folder/
+    # 653fab6  -- [HEAD]    feat: add compadd hook (4 hours ago)
     # word - the string to be inserted
+    # example word:
+    # Documents/Folder/
+    # 653fab6
     local dscr word i
+    
     for i in {1..$#__hits}; do
         word=$__hits[i] dscr=$__dscr[i]
         if [[ -n $dscr ]]; then
             dscr=${dscr//$'\n'}
         elif [[ -n $word ]]; then
             dscr=$word
         fi
+
+        # Hook defined by user to alter the description of the completion
+        fzf_tab_compadd_hook
```

Comment:
Please don't call `fzf_tab_compadd_hook` in the loop, this is very low inefficient.

Just put it before this loop. And check whether it exists before calling it (You can use `$+functions[fzf_tab_compadd_hook]` here).

Reply:
The issue when doing this was that when out of the loop I didnt managed to get the original hits + description pair

for example
```
fzf_tab_compadd_hook() {
  if [[ $_fzf_tab_curcontext == *systemctl-* ]]; then
      __dscr=()
      for service in $__hits; do
	      if [[ $(systemctl status $service) == *inactive* ]]; then
		      __dscr+=("[X]$service")
  		  else
			  __dscr+=("[O]$service")
   		  fi
      done
  fi
}
```
For a systemctl status it work but when the description is bigger (for a commit) this snippet loose the original description.

Reply:
When the hook is above the loop here is what I was using, but loosing the completion. And I didnt found a way to make it batch well at the moment and ended up with another loop
```zsh
fzf_tab_compadd_hook() {

    # if [[ -n $isfile ]]; then
    #   __dscr=(${(f)"$(devicon-lookup <<< ${(pj:\n:)__hits})"})
    # fi
    case $_fzf_tab_curcontext in
      (*)
        __dscr=()
        for hit in $__hits; do
          case $group in
            (function) icon="ƒ" ;;
            (alias) icon="Ą" ;;
            (*branch*) icon="" ;;
            (*command*) icon="" ;;
            (*file*) icon=$(echo $hit | devicon-lookup) hit="" ;;
            (*target*) icon="" ;;
            (*option*)
              case $hit in
                (*rm*|*delete**remove*)
                  icon=" " ;;
                (*help) icon="" ;; # icon=" " icon=" " icon=""
                (*) icon="" ;; # icon=" "
              esac
              ;;
            (*) icon=" " ;;
          esac

          __dscr+=("$icon $hit")
        done
        ;;
    esac
  }

```

Reply:
> For a systemctl status it work but when the description is bigger (for a commit) this snippet loose the original description.

My code is just a demo. In your case, you should edit the description in place.

```
for i in {1..$#__hits}; do
    __dscr[i]=$SOME_ICON$__dscr[i]
done
```

Reply:
I dont even get ho to assign a value to __dscr[i] without fucking it up
```zsh
fzf_tab_compadd_hook() {
  SOME_ICON="ƒ"
  local i
  for i in {1..$#__hits}; do
    # __dscr[i]=$SOME_ICON$__dscr[i]
    # __dscr[i]=${__dscr[i]}
    # $__dscr[i]="hello"
    # ${__dscr[i]}="hello"
    # __dscr[i]="hello"
    # __dscr[i]="hello "
    # __dscr[i]="hello \n"
  done
}
```

Reply:
Sometimes `$__dscr` is empty (eg. file name has no description). In this case, you should use `$__hits` instead.

```zsh
if (( $#__dscr )); then
    for i in {1..$#__hits}; do
        __dscr[i]="[O] "$__dscr[i]
    done
else
    for i in {1..$#__hits}; do
        __dscr[i]="[O] "$__hits[i]
    done
fi
```